File: src/cse/Environment.java
package cse;


import cse.ele.EleValueOrTuple;

import java.util.HashMap;

/**
 * Environment to keep entries on names and their values.
 */

public class Environment {
    private Environment parent;
    private final HashMap<String, EleValueOrTuple> memory;

    /**
     * primary environment.
     */
    Environment() {
        this.memory = new HashMap<>();
        remember("Print", null);
        remember("Isstring", null);
        remember("Isinteger", null);
        remember("Istruthvalue", null);
        remember("Istuple", null);
        remember("Isfunction", null);
        remember("Null", null);
        remember("Order", null);
        remember("Stern", null);
        remember("Stem", null);
        remember("ItoS", null);
        remember("neg", null);
        remember("not", null);
        remember("Conc", null);
    }

    /**
     * empty sub environment.
     */
    Environment(Environment parent) {
        this.memory = new HashMap<>();
        this.parent = parent;
    }

    /**
     * sub environment with one entry.
     */
    Environment(Environment parent, String key, EleValueOrTuple value) {
        this.memory = new HashMap<>();
        this.parent = parent;
        remember(key, value);
    }

    /**
     * Remember an entry. Error if already defined.
     */
    void remember(String key, EleValueOrTuple value) {
        if (memory.containsKey(key)) {
            throw new RuntimeException("Variable is already defined: " + key);
        }
        memory.put(key, value);
    }

    /**
     * Get the value of a variable.
     * Returns null if defined in primary environment.
     * Throws error if undefined.
     */
    EleValueOrTuple lookup(String id) {
        if (memory.containsKey(id)) {
            return memory.get(id);
        }
        if (parent == null) {
            // Primary Environment and not found
            throw new RuntimeException("Undefined variable: " + id);
        }
        return parent.lookup(id);
    }

    @Override
    public String toString() {
        if (parent != null) {
            String[] data = new String[memory.size()];
            int i = 0;
            for (String key : memory.keySet()) {
                data[i++] = "[" + memory.get(key) + "/" + key + "]";
            }
            return parent + " > " + String.join("", data);
        }
        return "PE";
    }
}

File: reCreate/main/CSE/Environment.py
from .elements.EleValueOrTuple import EleValueOrTuple
class Environment:
    def __init__(self, parent=None, key=None, value=None):
        self.parent = parent
        self.memory = {}
            
        if key is None and value is None:
            self.initializePrimaryEnv()
        elif key and value:
            self.remember(key, value)

    def initializePrimaryEnv(self):
        self.remember("Print", None)
        self.remember("Isstring", None)
        self.remember("Isinteger", None)
        self.remember("Istruthvalue", None)
        self.remember("Istuple", None)
        self.remember("Isfunction", None)
        self.remember("Null", None)
        self.remember("Order", None)
        self.remember("Stern", None)
        self.remember("Stem", None)
        self.remember("ItoS", None)
        self.remember("neg", None)
        self.remember("not", None)
        self.remember("Conc", None)

    def remember(self, key, value):
        if key in self.memory:
            raise RuntimeError("Variable is already defined: " + key)
        self.memory[key] = value

    def lookup(self, id):
        if id in self.memory:
            return self.memory[id]
        if self.parent is None:
            raise RuntimeError("Undefined variable: " + id)
        return self.parent.lookup(id)

    def toString(self):
        if self.parent is not None:
            data = [f"[{self.memory[key]}/{key}]" for key in self.memory]
            return str(self.parent) + " > " + "".join(data)
        return "PE"
